### ISSUE TITLE
Single step study search studies with full parameter

### ISSUE THREAD

**Issue by cpommier** [plantbreeding/API#366]
The call API/Specification/Studies/Studies_GET.yaml is important and must remain in futur version as a single step call, not the two step system used for POST /search/studies which returns a searchToken.

Two points:

- All search criterions
the following field shouldn't be deprecated:

        description: DEPRECATED in v1.3 - see /search/studies
        name: germplasmDbIds

        description: DEPRECATED in v1.3 - see /search/studies
        name: observationVariableDbIds

- GET & POST:
we should have both
GET /studies response study object 
and
POST /studies response study object

**Comment by cpommier** [comment]
After Discussion with Peter, the second point is on hold. The first point seems still valid

**Comment by cpommier** [comment]
There are some feedbacks regarding the two step query approach, not too enthusiastic. Feel free to add items below.
The understanding is that the server side search object can live for some time (minutes ?). It can be simply the list of search criterions received from the POST.

## PROS

- Speed improvement with prepared SQL queries. This implies long connection with the DB which is not necessarily a good practice.

## CONS

- Risk of inconsistency between the server side and client side search criterions. To avoid this, each request should be systematically sent with two steps. 
- More complex more things to check both on client and server side, hence more risk of bugs or instability 
- Obsolescence of the token which has to be handled on both side. It can furthermore totally fail if there is a server side restart of the service or the session.

**Comment by nickmorales** [comment]
I think Cyril's points are valid. If a client's previous search POST is not valid on the server anymore (because of a timeout or other server-side mechanism for handling these cached search parameters), is the client supposed to make a new POST immediately followed by a new GET?

**Comment by dauglyon** [comment]
Two approaches for the problem of persistence on the data side:

1) This one requires changing the BrAPI standard. Perhaps a solution to this problem is to allow the POST call to act like `/allematrices-search`. It can either return the full response, or an asyncStatus object.

2) This one can be implemented under the current standard. If the query is small enough to be contained in the GET URL parameters, I might suggest the server return a URL safe identifier encoding all nesscary search params to pass to the GET request. This requires no persistance of the request on the server side. EG:

- Client POSTs `http://.../search/entity` with body `Body:{germplasmDbIds:["14"]}`
- Server losslessly encodes `{germplasmDbIds:["14"]}` to `e2dlcm1wbGFzbURiSWRzOlsiMTQiXX0=`
- Server responds `Body: {searchId:"e2dlcm1wbGFzbURiSWRzOlsiMTQiXX0="}`
- Client GETs `http://.../search/entity/e2dlcm1wbGFzbURiSWRzOlsiMTQiXX0=`
- Server decodes `e2dlcm1wbGFzbURiSWRzOlsiMTQiXX0=` to `{germplasmDbIds:["14"]}` and performs search.
- Server responds with search results

(HTTP 413 could be used to reject POST bodies which cannot fit in a url if this is the only search method implemented)

**Comment by cpommier** [comment]
The more I think about this, the more I think me must ensure that those request remain stateless to be fully RESTful. Two step query address a specific use case that don't need to be the general approach. We must make the life of client developers as easy as possible. 
Some comments
Solution 2: having an ID that is in fact not an ID but an encoded query string will be both confusing and problematic. Things should be explicit. I strongly advocate against this one.

Solution 1: 
If the return type changes, ie it returns either an asynchStatus object or an array, it can be really disturbing for the client. We could have a composite return type, with a nullable search token in the response body alongside the full response that could be used as a search parameter. But that's complicated and possibly misleading.

The most explicit way would be in BrAPI v2 to have all of the following calls : 
```
  GET /entity?param=value...
      Returns -> A list of "entities" filtered by the parameters
    
    POST /search/entity/json
      Body -> {"param": ["value1", "value2"...]}
      Returns -> The JSON response structure, with pagination and the list of "entities" filtered by the parameters
    
    POST /search/entity/tab
      Body -> {"param": ["value1", "value2"...]}
      Returns -> The tab (CSV, TSV) response structure, with pagination and the list of "entities" filtered by the parameters

    POST /search/entity/tab-file
      Body -> {"param": ["value1", "value2"...]}
      Returns -> The tab (CSV, TSV) file with no pagination and the link to a generated file containing the list of "entities" filtered by the parameters

    POST /search/entity/
      Body -> {"param": ["value1", "value2"...]}
      Returns -> A search results DbId to be used to retrieve the results
    
    GET /search/entity/{searchResultsDbId}
      Returns -> A list of "entities" filtered by the parameters in the related POST call
```

**Comment by BrapiCoordinatorSelby** [comment]
My apologies, I should have been following this discussion more closely and responded sooner. I'd like to make some clarifying points, starting from the top of the thread. 

>@cpommier All search criterions
the following field shouldn't be deprecated: ... germplasmDbIds ... observationVariableDbIds

The reason these were deprecated is because they were defined as lists. There are some technical issues with allowing arbitrarily long lists as parameters in GET requests, URL length being the primary problem. I am happy to bring these criterion back, but as single values only. 

>@cpommier The understanding is that the server side search object can live for some time (minutes ?). It can be simply the list of search criterions received from the POST.

I think there is a miss-understanding here. The search object should be treated as an entity and given space in the database. By default, it should live forever. if database space is limited, each search entity in the database could be given an expiration and removed by an automated process, but the intention here would be that expiration would be on the order of months or years. Assuming the average search record in a DB is 200 bytes, 5 million searches POSTed would only be 1GB. 

> @cpommier This implies long connection with the DB which is not necessarily a good practice.

This is not the intended implication, see above.

>@cpommier Risk of inconsistency between the server side and client side search criterions.

When the client creates a search object and POSTs it, those search criteria are "locked in" and given an ID (searchResultsDbId). The client can refer to the results of those locked search criteria as many times as needed using the ID. If the client has new search criteria, a new search object needs to be created.

> @cpommier Obsolescence of the token which has to be handled on both side. It can furthermore totally fail if there is a server side restart of the service or the session.

This will not happen if the database is used, see above.

> @nickmorales is the client supposed to make a new POST immediately followed by a new GET?

Yes. This seems like an unlikely scenario if the search request object is treated like an entity and stored

>@dauglyon Two approaches for the problem of persistence on the data side:

@dauglyon sorry, but I agree with @cpommier on both your solutions. (1) could be problematic, the only reason `allelematrix-search` works is that very few people know that it can technically work that way. (2) is an overly complex solution to a much simpler problem

> @cpommier The more I think about this, the more I think me must ensure that those request remain stateless to be fully RESTful.... We must make the life of client developers as easy as possible.

This approach is stateless and fully RESTful. There is no session or history being maintained between the client and server. Search objects are being treated as entities, the client POSTs a new entity to the server, then retrieves some calculated information related to that entity. Just like you might POST a new Germplasm object, then retrieve its calculated pedigree information. The life of the client developer is not made difficult, all of the complexity here lives on the server and the ability to record search requests. The intention was to give the client dev MORE power and flexibility.

> @cpommier Two step query address a specific use case that don't need to be the general approach.

This is a totally valid opinion and I see your point. I made an executive decision around new searching approach which was obviously not fully accepted by the community. And, as is my nature, I may have over engineered a solution to a problem that didn't exist. I am sorry for this, I have learned from the experience, and I will take steps to make sure something similar doesn't happen in the future.

That said, I still believe I made the right call, and I would like to quickly layout my decision making process. First, we needed to separate out and standardize the `GET /entity` simple filter requests. Next, the `POST /entity-search` calls needed to be re-arranged to be more RESTful. This was more or less a style choice rather than a technical one, but it is a common solution in other RESTful APIs. 
Now, the most common use case I had seen for using the `POST /enitity-search` call was for sending long lists of DbIds and retrieving all the entity details for that long list. This was completely anecdotal, I had not conducted any surveys, but my impression from talking to people was this is a very common use case. The current design tried to optimize this sort of scenario by allowing the client to send that large request only once, then page through the results as necessary, referencing the search object by the much shorter ID. This has the added bonus that if the server were to keep track of the results of a search, not just the request, then large, complex queries only need to be run once, instead of rerunning the query for each page of results. For smaller, simpler requests, I felt that 2 API calls instead of 1 was not a significant burden on performance. 
There are also several outstanding issues which address things like sending complex custom search logic and wild cards as part of a search request. It is my hope that this separation of request and response can act as a model while we try to tackle even more complicated searching requests. 

> @cpommier The most explicit way would be in BrAPI v2 to have all of the following calls 

In this set of calls you are also addressing the return of different Content-Types. This response is already too long, so I'm going to leave that discussion for another thread. However, I will caution you on having too many ways to do the same thing. (https://github.com/plantbreeding/API/issues/151#issuecomment-363812697) A POSTed search request should either return an object, or a results ID. Having both options will mean client applications will have to handle both and test which one a particular server supports.

Again, sorry this is so long and late. @cpommier please feel free to email me and we can work with your team to brainstorm solutions. Everything we have been discussing here has been theoretical, I would love to see some problematic use cases in practice.

**Comment by BrapiCoordinatorSelby** [comment]
This issue was discussed by the BrAPI review board and a decision was reached. A server may chose to implement single step POST searches or asynchronous multi-step POST/GET searches. The client is responsible for reading the HTTP status code (which they should be doing anyway, 200 vs 400 etc) to determine the correct action to retrieve the final data results. This way the server can chose how to best tune complex searching to their environment and the client has a standard algorithm which will work for all server configurations.
